### PR TITLE
Even better(er) binary quantization

### PR DESCRIPTION
**Pull request by benwtrent**
This measurably improves BBQ by adjusting the underlying algorithm to an optimized per vector scalar quantization.

This is a brand new way to quantize vectors. Instead of there being a global set of upper and lower quantile bands, these are optimized and calculated per individual vector. Additionally, vectors are centered on a common centroid. 

This allows for an almost 32x reduction in memory, and even better recall than before at the cost of slightly increasing indexing time.

Additionally, this new approach is easily generalizable to various other bit sizes (e.g. 2 bits, etc.). While not taken advantage of yet, we may update our scalar quantized indices in the future to use this new algorithm, giving significant boosts in recall.

The recall gains spread from 2% to almost 10% for certain datasets with an additional 5-10% indexing cost when indexing with HNSW when compared with current BBQ.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatVectorsScorer.java [hunk 1]
```diff
@@ -48,13 +48,8 @@ class ES816BinaryFlatVectorsScorer implements FlatVectorsScorer {
     public RandomVectorScorerSupplier getRandomVectorScorerSupplier(
         VectorSimilarityFunction similarityFunction,
         KnnVectorValues vectorValues
-    ) throws IOException {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatVectorsScorer.java:51]: removing write path for old codec

File: server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatVectorsScorer.java [hunk 2]
```diff
@@ -90,61 +85,11 @@ public RandomVectorScorer getRandomVectorScorer(
...
     }
 
-    RandomVectorScorerSupplier getRandomVectorScorerSupplier(
-        VectorSimilarityFunction similarityFunction,
-        ES816BinaryQuantizedVectorsWriter.OffHeapBinarizedQueryVectorValues scoringVectors,
-        BinarizedByteVectorValues targetVectors
-    ) {
-        return new BinarizedRandomVectorScorerSupplier(scoringVectors, targetVectors, similarityFunction);
-    }
-
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatVectorsScorer.java:100]: removing write path for old codec

File: server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatVectorsScorer.java [hunk 3]
```diff
@@ -90,61 +85,11 @@ public RandomVectorScorer getRandomVectorScorer(
...
     @Override
     public String toString() {
         return "ES816BinaryFlatVectorsScorer(nonQuantizedDelegate=" + nonQuantizedDelegate + ")";
     }
 
-    /** Vector scorer supplier over binarized vector values */
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatVectorsScorer.java:106]: removing write path for old codec

File: server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryQuantizedVectorsFormat.java [hunk 4]
```diff
@@ -62,7 +62,7 @@ public ES816BinaryQuantizedVectorsFormat() {
 
     @Override
     public FlatVectorsWriter fieldsWriter(SegmentWriteState state) throws IOException {
-        return new ES816BinaryQuantizedVectorsWriter(scorer, rawVectorFormat.fieldsWriter(state), state);
+        throw new UnsupportedOperationException();
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryQuantizedVectorsFormat.java:65]: removing write path for old codec

File: server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816HnswBinaryQuantizedVectorsFormat.java [hunk 5]
```diff
@@ -25,10 +25,8 @@
 import org.apache.lucene.codecs.hnsw.FlatVectorsFormat;
 import org.apache.lucene.codecs.lucene99.Lucene99HnswVectorsFormat;
 import org.apache.lucene.codecs.lucene99.Lucene99HnswVectorsReader;
-import org.apache.lucene.codecs.lucene99.Lucene99HnswVectorsWriter;
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/es816/ES816HnswBinaryQuantizedVectorsFormat.java:28]: removing write path for old codec

File: server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatRWVectorsScorer.java [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+import org.elasticsearch.simdvec.ESVectorUtil;
+
+import java.io.IOException;
+
+import static org.apache.lucene.index.VectorSimilarityFunction.COSINE;
+import static org.apache.lucene.index.VectorSimilarityFunction.EUCLIDEAN;
+import static org.apache.lucene.index.VectorSimilarityFunction.MAXIMUM_INNER_PRODUCT;
+
+/** Vector scorer over binarized vector values */
+class ES816BinaryFlatRWVectorsScorer implements FlatVectorsScorer {
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatRWVectorsScorer.java:40]: For old codec backward compat testing. Copy paste of old scorer, but now allowing write path

File: server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatVectorsScorerTests.java [hunk 7]
```diff
@@ -59,7 +59,7 @@ public void testScore() throws IOException {
         short quantizedSum = (short) random().nextInt(0, 4097);
         float normVmC = random().nextFloat(-1000f, 1000f);
         float vDotC = random().nextFloat(-1000f, 1000f);
-        ES816BinaryFlatVectorsScorer.BinaryQueryVector queryVector = new ES816BinaryFlatVectorsScorer.BinaryQueryVector(
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryFlatVectorsScorerTests.java:62]: For old codec backward compat testing

File: server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryQuantizedRWVectorsFormat.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.lucene.codecs.hnsw.FlatVectorsWriter;
+import org.apache.lucene.codecs.lucene99.Lucene99FlatVectorsFormat;
+import org.apache.lucene.index.SegmentWriteState;
+
+import java.io.IOException;
+
+/**
+ * Copied from Lucene, replace with Lucene's implementation sometime after Lucene 10
+ */
+public class ES816BinaryQuantizedRWVectorsFormat extends ES816BinaryQuantizedVectorsFormat {
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryQuantizedRWVectorsFormat.java:33]: For old codec backward compat testing

File: server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryQuantizedVectorsFormatTests.java [hunk 9]
```diff
@@ -63,7 +63,7 @@ protected Codec getCodec() {
         return new Lucene100Codec() {
             @Override
             public KnnVectorsFormat getKnnVectorsFormatForField(String field) {
-                return new ES816BinaryQuantizedVectorsFormat();
+                return new ES816BinaryQuantizedRWVectorsFormat();
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryQuantizedVectorsFormatTests.java:66]: For old codec backward compat testing

File: server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryQuantizedVectorsWriter.java [hunk 10]
```diff
@@ -77,7 +77,7 @@ class ES816BinaryQuantizedVectorsWriter extends FlatVectorsWriter {
     private final List<FieldWriter> fields = new ArrayList<>();
     private final IndexOutput meta, binarizedVectorData;
     private final FlatVectorsWriter rawVectorDelegate;
-    private final ES816BinaryFlatVectorsScorer vectorsScorer;
+    private final ES816BinaryFlatRWVectorsScorer vectorsScorer;
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816BinaryQuantizedVectorsWriter.java:80]: For old codec backward compat testing

File: server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816HnswBinaryQuantizedRWVectorsFormat.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.lucene.index.SegmentWriteState;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+
+import static org.apache.lucene.codecs.lucene99.Lucene99HnswVectorsFormat.DEFAULT_BEAM_WIDTH;
+import static org.apache.lucene.codecs.lucene99.Lucene99HnswVectorsFormat.DEFAULT_MAX_CONN;
+import static org.apache.lucene.codecs.lucene99.Lucene99HnswVectorsFormat.DEFAULT_NUM_MERGE_WORKER;
+
+class ES816HnswBinaryQuantizedRWVectorsFormat extends ES816HnswBinaryQuantizedVectorsFormat {
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816HnswBinaryQuantizedRWVectorsFormat.java:34]: For old codec backward compat testing

File: server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816HnswBinaryQuantizedVectorsFormatTests.java [hunk 12]
```diff
@@ -59,7 +59,7 @@ protected Codec getCodec() {
         return new Lucene100Codec() {
             @Override
             public KnnVectorsFormat getKnnVectorsFormatForField(String field) {
-                return new ES816HnswBinaryQuantizedVectorsFormat();
+                return new ES816HnswBinaryQuantizedRWVectorsFormat();
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/index/codec/vectors/es816/ES816HnswBinaryQuantizedVectorsFormatTests.java:62]: For old codec backward compat testing

**Review by john-wagster — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/codec/vectors/es818/OptimizedScalarQuantizer.java [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+    private double loss(float[] vector, float[] interval, int points, float norm2) {
+        double a = interval[0];
+        double b = interval[1];
+        double step = ((b - a) / (points - 1.0F));
+        double stepInv = 1.0 / step;
+        double xe = 0.0;
+        double e = 0.0;
+        for (double xi : vector) {
+            double xiq = (a + step * Math.round((clamp(xi, a, b) - a) * stepInv));
```
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/es818/OptimizedScalarQuantizer.java:156]: yeap, this seems about right 😅

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/codec/vectors/es818/OptimizedScalarQuantizer.java:156]: LOL, I should indeed add some comments here. We are basically calculating the error of quantizing and then unquantizing and shifting up or down based on the mis-calculation.

**Review by pmpailis — COMMENTED** [review]
Reviewed tests and the main set of changes against the `816*` versions and they LGTM. The math changes are a bit more difficult to review; but will give it a more thorough go on Monday 😅 (no need to wait though; please feel free to proceed w/o waiting for my review on that part)

**Comment by benwtrent** [comment]
> It would be nice to add some documentation to the format: it looks like the queries are still 4 bits quantized?

The on disk format is very similar. I can add some docs on that to the format

**Review by mayya-sharipova — APPROVED** [review]
@benwtrent Thanks Ben! Great work!

Similar to Panos, I reviewed file formats, as don't follow all the math in quantizer but I trust you and Tom got it right.

**Pull request merged**